lib/ecto/query/builder/select.ex

```diff
@@ -41,6 +41,20 @@ defmodule Ecto.Query.Builder.Select do
     {expr, params}
   end
 
+  # Map
+  defp escape({:%{}, _, pairs}, params, vars) do
+    {pairs, params} = Enum.map_reduce pairs, params, fn({k, v}, acc) ->
+      unless is_atom(k) or is_binary(k) or is_number(k) do
+        raise ArgumentError, "keys can only be atoms/binaries/integers/floats"
```

Comment:
Sorry, I think there is a Builder.error! "..." that would be more appropriate to raise as it contains more info.